frontend/packages/operator-lifecycle-manager/src/components/create-operand.tsx

```diff
@@ -104,297 +178,465 @@ const defaultValueFor = (field: OperandField) => {
+            flattenNestedProperties(nestedProperty, nestedPropertyName, providedAPI, obj, {
+              currentCapabilities: [
+                ...currentCapabilities,
+                `${SpecCapability.arrayFieldGroup}${name}`,
               ],
+              currentPath: [...currentPath, `${name}[${index}]`], // Array field paths must include an index
+              fields,
+              required: (property?.required || []).includes(nestedPropertyName),
             }),
-          );
-        case undefined:
-          return null;
-        default:
-          return [
-            {
-              path: key,
-              displayName: _.startCase(key),
-              type: val.type,
-              required: _.get(openAPI.properties.spec, 'required', []).includes(key),
-              validation: _.pick(val, [...Object.keys(Validations)]),
-              capabilities: capabilitiesFor(val),
-            },
-          ];
+          ),
+        2,
+      );
+
+    // This property is not an array or object, so it can be mapped to a specific descriptor
+    default:
+      const path = [...currentPath, name].join('.');
+
+      // ProvidedAPI should only have a single descriptor for each array field. Regardless of the
+      // index of this field, use the providedAPI.specDescriptor at index 0.
+      const providedAPIField = _.find<CRDDescription>(providedAPI?.specDescriptors, {
+        path: modifyArrayFieldPathIndex(path, () => 0),
+      });
+      return [
+        ...fields,
+        {
+          capabilities: _.union(
+            providedAPIField?.['x-descriptors'] || [],
+            currentCapabilities,
+            capabilitiesFor(property),
+          ),
+          description: providedAPIField?.description || property.description,
+          displayName: providedAPIField?.displayName || _.startCase(name),
+          path,
+          required,
+          type: property.type,
+          validation: _.pick(property, Object.keys(Validations)),
+        },
+      ];
+  }
+};
+
+// Returns traversal depth of an OpenAPI spec property.
+const getPropertyDepth = (property, depth = 0) => {
+  // If this property is not an array or object, we have reached the maximum depth
+  if (!property || !['object', 'array'].includes(property.type)) {
+    return depth;
+  }
+
+  // Return the maximum depth of the nested properties
+  return Math.max(
+    0,
+    ..._.map(property?.properties || property?.items?.properties, (nestedProperty) =>
+      getPropertyDepth(nestedProperty, depth + 1),
+    ),
+  );
+};
+
+/*
+ * Maps openAPI spec properties to OperandField[] array. This will return all fields with a traversal
+ * depth less than the 'depth' argument, and will combine matching providedAPI descriptors with the
+ * corresponding OpenAPI spec property in a way that providedAPI data will take precedence over
+ * openAPI spec properties.
+ */
+const fieldsForOpenAPI = (openAPI, providedAPI, obj, depth = MAX_DEPTH) => {
```

Comment:
Changed the way openAPI swagger definitions are built. They now take into account the 'buffer' or existing data that was synced from the YAML editor. We have to duplicate arrayFieldGroups so that enough fields will exist on the form to hold the values from synced data.

Reply:
I also pass in the providedAPI prop so that we can use the descriptors in those to override the api definition. So, providedAPI always takes precedence.